python/rikai/contrib/torch/inspect/ssd.py

```diff
@@ -0,0 +1,191 @@
+        features = self.backend.backbone(images.tensors)
+        features = list(features.values())
+        head_outputs = self.backend.head(features)
+        pred_anchors = self.backend.anchor_generator(images, features)
+
+        scores = head_outputs["cls_logits"]
+        bbox_regression = head_outputs["bbox_regression"]
+        pred_scores = F.softmax(scores, dim=-1)
+
+        num_classes = pred_scores.size(-1)
+        device = pred_scores.device
+
+        detections: List[Dict[str, Tensor]] = []
+        for boxes, scores, anchors, image_shape in zip(
+            bbox_regression, pred_scores, pred_anchors, images.image_sizes
+        ):
+            boxes = self.backend.box_coder.decode_single(boxes, anchors)
+            boxes = clip_boxes_to_image(boxes, image_shape)
+
+            image_boxes = []
+            image_scores = []
+            image_labels = []
+            image_all_scores = []
+            for label in range(1, num_classes):
+                score = scores[:, label]
+
+                keep_idxs = score > self.backend.score_thresh
+                score = score[keep_idxs]
+                all_scores = scores[keep_idxs]
+                box = boxes[keep_idxs]
+
+                # keep only topk scoring predictions
+                num_topk = min(self.backend.topk_candidates, score.size(0))
+                score, idxs = score.topk(num_topk)
+                box = box[idxs]
+                all_scores = all_scores[idxs]
+
+                image_boxes.append(box)
+                image_scores.append(score)
+                image_all_scores.append(all_scores)
+                image_labels.append(
+                    torch.full_like(
+                        score,
+                        fill_value=label,
+                        dtype=torch.int64,
+                        device=device,
+                    )
+                )
+
+            image_boxes = torch.cat(image_boxes, dim=0)
+            image_scores = torch.cat(image_scores, dim=0)
+            image_labels = torch.cat(image_labels, dim=0)
+            image_all_scores = torch.cat(image_all_scores, dim=0)
+
+            # non-maximum suppression
+            keep = batched_nms(
+                image_boxes,
+                image_scores,
+                image_labels,
+                self.backend.nms_thresh,
+            )
+            keep = keep[: self.backend.detections_per_img]
+            top_scores, idx = image_all_scores[keep, 1:].topk(2)
+            detections.append(
+                {
+                    "boxes": image_boxes[keep],
+                    "scores": top_scores,
+                    "labels": idx + 1,
+                }
+            )
+        detections = self.backend.transform.postprocess(
+            detections, images.image_sizes, original_image_sizes
+        )
+        return detections
+
+
+class_scores_extractor_pre_processing = pre_processing
+
+
+def class_scores_extractor_post_processing(
```

Comment:
confusing to have transforms spread across multiple folders/files

Reply:
My intention is that the pre/post processing should be close to the model itself. 

For ssd / faster-rcnn and etc, the model code does not exist in rikai repo, as they are in torchvision. so thats where the transforms package is.

Reply:
sounds like you're leaning towards organizing by module rather than by functionality then right?

Reply:
yes, lets organize these code by module.